test/Kestrel.InMemory.FunctionalTests/Http2/Http2ConnectionTests.cs

```diff
@@ -3786,6 +3793,262 @@ public void UnexpectedExceptionDuringFrameProcessingLoggedAWarning()
+                new KeyValuePair<string, string>(HeaderNames.Path, "/"),
+                new KeyValuePair<string, string>(HeaderNames.Scheme, "http"),
+            };
+            await InitializeConnectionAsync(_appAbort);
+
+            await StartStreamAsync(1, headers, endStream: false);
+
+            await WaitForStreamErrorAsync(1, Http2ErrorCode.INTERNAL_ERROR, "The connection was aborted by the application.");
+
+            // There's a race when the appfunc is exiting about how soon it unregisters the stream.
+            for (var i = 0; i < 10; i++)
+            {
+                await SendDataAsync(1, new byte[100], endStream: false);
+            }
+            await SendRstStreamAsync(1);
+
+            // Send an extra frame to make it fail
+            switch (finalFrameType)
+            {
+                case Http2FrameType.DATA:
+                    await SendDataAsync(1, new byte[100], endStream: true);
+                    break;
+
+                case Http2FrameType.HEADERS:
+                    await SendHeadersAsync(1, Http2HeadersFrameFlags.END_STREAM | Http2HeadersFrameFlags.END_HEADERS, _requestTrailers);
+                    break;
+
+                default:
+                    throw new NotImplementedException(finalFrameType.ToString());
+            }
+
+            await WaitForConnectionErrorAsync<Http2ConnectionErrorException>(ignoreNonGoAwayFrames: false, expectedLastStreamId: 1, Http2ErrorCode.STREAM_CLOSED,
+                CoreStrings.FormatHttp2ErrorStreamClosed(finalFrameType, 1));
+        }
+
+        [Theory]
+        [InlineData(Http2FrameType.DATA)]
+        [InlineData(Http2FrameType.HEADERS)]
+        public async Task AbortedStream_ResetsAndDrainsRequest_RefusesFramesAfterCooldownExpires(Http2FrameType finalFrameType)
+        {
+            var mockSystemClock = new MockSystemClock();
+            _connectionContext.ServiceContext.SystemClock = mockSystemClock;
+
+            var headers = new[]
+            {
+                new KeyValuePair<string, string>(HeaderNames.Method, "POST"),
+                new KeyValuePair<string, string>(HeaderNames.Path, "/"),
+                new KeyValuePair<string, string>(HeaderNames.Scheme, "http"),
+            };
+            await InitializeConnectionAsync(_appAbort);
+
+            await StartStreamAsync(1, headers, endStream: false);
+
+            await WaitForStreamErrorAsync(1, Http2ErrorCode.INTERNAL_ERROR, "The connection was aborted by the application.");
+
+            // There's a race when the appfunc is exiting about how soon it unregisters the stream.
+            for (var i = 0; i < 10; i++)
+            {
+                await SendDataAsync(1, new byte[100], endStream: false);
+            }
+
+            // Timeouts already offset by one Heartbeat Interval.
+            mockSystemClock.UtcNow += Constants.RequestBodyDrainTimeout + TimeSpan.FromMilliseconds(1);
```

Comment:
Generally I prefer testing that the error doesn't occur at exactly the timeout, and does occur one tick after. The tick vs. millisecond isn't a big deal, but testing that the error doesn't occur right before the timeout fires has caught an issue where I was using the wrong timeout before.